File: 1182-shortest-distance-to-target-color/1182-shortest-distance-to-target-color.py
class Solution:
    def shortestDistanceColor(self, colors: List[int], queries: List[List[int]]) -> List[int]:
        n = len(colors)
        dis = [[inf] * 3 for _ in range(n)]
        ci = [inf] * 3
        for i in range(n):
            ci[colors[i] - 1] = i
            for c in range(3):
                if ci[c] != inf:
                    dis[i][c] = i - ci[c]
        ci = [inf] * 3
        for i in reversed(range(n)):
            ci[colors[i] - 1] = i
            for c in range(3):
                if ci[c] != inf:
                    dis[i][c] = min(dis[i][c], ci[c] - i)
        return [dis[i][c - 1] if dis[i][c - 1] != inf else -1 for i, c in queries]

File: 1182-shortest-distance-to-target-color/1182-shortest-distance-to-target-color.java
class Solution {
    public List<Integer> shortestDistanceColor(int[] colors, int[][] queries) {
        ArrayList<Integer>[] colorIndices = new ArrayList[]{
            new ArrayList<Integer>(),
            new ArrayList<Integer>(),
            new ArrayList<Integer>()
        };
        for(int i = 0; i < colors.length; i++)
            colorIndices[colors[i] - 1].add(i);
        
        List<Integer> ans = new ArrayList<>();
        for(int i = 0; i < queries.length; i++){
            int idx = queries[i][0], color = queries[i][1];
            ArrayList<Integer> indices = colorIndices[color - 1];
            if(indices.size() == 0){
                ans.add(-1);
                continue;
            }
            
            int lo = 0, hi = indices.size() - 1;
            while(lo < hi){
                int mid = lo + (hi - lo) / 2;
                if(indices.get(mid) >= idx)
                    hi = mid;
                else
                    lo = mid + 1;
            }
            
            ans.add(
                Math.min(
                    Math.abs(indices.get(Math.max(lo - 1, 0)) - idx),
                    Math.abs(indices.get(lo) - idx)
                )
            );
        }
        return ans;
    }
}